va/dns_account_test.go

```diff
@@ -0,0 +1,164 @@
+		" (account: " + testAccountURI + ")"
+	test.AssertEquals(t, prob.String(), expectedErr)
+}
+
+func TestDNSAccount01ValidationIP(t *testing.T) {
+	va, _ := setup(nil, "", nil, nil)
+
+	_, err := va.validateDNSAccount01(ctx, identifier.NewIP(netip.MustParseAddr("127.0.0.1")), expectedKeyAuthorization, testAccountURI)
+	prob := detailedError(err)
+
+	test.AssertEquals(t, prob.Type, probs.MalformedProblem)
+}
+
+func TestDNSAccount01ValidationInvalid(t *testing.T) {
+	var notDNS = identifier.ACMEIdentifier{
+		Type:  identifier.IdentifierType("iris"),
+		Value: "790DB180-A274-47A4-855F-31C428CB1072",
+	}
+
+	va, _ := setup(nil, "", nil, nil)
+
+	_, err := va.validateDNSAccount01(ctx, notDNS, expectedKeyAuthorization, testAccountURI)
+	prob := detailedError(err)
+
+	test.AssertEquals(t, prob.Type, probs.MalformedProblem)
+}
+
+func TestDNSAccount01ValidationServFail(t *testing.T) {
+	va, _ := setup(nil, "", nil, nil)
+
+	_, err := va.validateDNSAccount01(ctx, identifier.NewDNS("servfail.com"), expectedKeyAuthorization, testAccountURI)
+
+	prob := detailedError(err)
+	test.AssertEquals(t, prob.Type, probs.DNSProblem)
+}
+
+func TestDNSAccount01ValidationNoServer(t *testing.T) {
+	va, log := setup(nil, "", nil, nil)
+	staticProvider, err := bdns.NewStaticProvider([]string{})
+	test.AssertNotError(t, err, "Couldn't make new static provider")
+
+	va.dnsClient = bdns.NewTest(
+		time.Second*5,
+		staticProvider,
+		metrics.NoopRegisterer,
+		clock.New(),
+		1,
+		"",
+		log,
+		nil)
+
+	_, err = va.validateDNSAccount01(ctx, identifier.NewDNS("localhost"), expectedKeyAuthorization, testAccountURI)
+	prob := detailedError(err)
+	test.AssertEquals(t, prob.Type, probs.DNSProblem)
+}
+
+func TestDNSAccount01ValidationOK(t *testing.T) {
+	va, _ := setup(nil, "", nil, nil)
+
+	_, prob := va.validateDNSAccount01(ctx, identifier.NewDNS("good-dns01.com"), expectedKeyAuthorization, testAccountURI)
+
+	test.Assert(t, prob == nil, "Should be valid.")
+}
+
+func TestDNSAccount01ValidationNoAuthorityOK(t *testing.T) {
+	va, _ := setup(nil, "", nil, nil)
+
+	_, prob := va.validateDNSAccount01(ctx, identifier.NewDNS("no-authority-dns01.com"), expectedKeyAuthorization, testAccountURI)
+
+	test.Assert(t, prob == nil, "Should be valid.")
+}
+
+func TestDNSAccount01ValidationEmptyAccountURI(t *testing.T) {
+	va, _ := setup(nil, "", nil, nil)
+
+	// The specific domain doesn't matter, as the function should
+	// reject the empty accountURI before DNS lookup.
+	ident := identifier.NewDNS("empty-uri-test.com")
+
+	// Call the validation function with an empty accountURI
+	_, err := va.validateDNSAccount01(ctx, ident, expectedKeyAuthorization, "")
+
+	// Assert that an error was returned
+	test.Assert(t, err != nil, "validateDNSAccount01 succeeded unexpectedly with an empty account URI")
+
+	// Assert the specific error type
+	prob := detailedError(err)
+	test.AssertEquals(t, prob.Type, probs.ConnectionProblem)
```

Comment:
When accountURI is missing in validateDNSAccount01, it triggers an InternalServerError. The test TestDNSAccount01ValidationEmptyAccountURI verifies that the resulting client-facing error is mapped to ConnectionProblem (via the WFE mapping). Is ConnectionProblem the most suitable error type for the client, or should we consider a more specific internal error, such as MissingParameter, which could map to a different client-facing error (e.g., Malformed) if needed in the future? However, since the accountURI originates from the JWS, this error scenario should theoretically never occur.

Reply:
ConnectionProblem is what we use as a catch-all for all 500s during validation, so it's fine to test for that here. But I think it would be even better to drop the call to detailedError, and instead directly assert that the returned error is of type berrors.ServerInternal.